arduino-ide-extension/src/browser/dialogs/ide-updater/ide-updater-dialog.tsx

```diff
@@ -143,22 +116,118 @@ export class IDEUpdaterDialog extends AbstractDialog<UpdateInfo> {
       Widget.detach(this.widget);
     }
     Widget.attach(this.widget, this.contentNode);
+    this.appendInitialButtons();
     super.onAfterAttach(msg);
-    this.update();
+  }
+
+  private clearButtons(): void {
+    while (this.controlPanel.firstChild) {
+      this.controlPanel.removeChild(this.controlPanel.firstChild);
+    }
+    this.closeButton = undefined;
+  }
+
+  private appendNotNowButton(): void {
+    this.appendCloseButton(
+      nls.localize('arduino/ide-updater/notNowButton', 'Not now')
+    );
+    if(this.closeButton) {
+      this.addCloseAction(this.closeButton, 'click')
+    }
+  }
+
+  private appendInitialButtons(): void {
+    this.clearButtons();
+
+    const skipVersionButton = this.createButton(
+      nls.localize('arduino/ide-updater/skipVersionButton', 'Skip Version')
+    );
+    skipVersionButton.classList.add('secondary');
+    skipVersionButton.classList.add('skip-version-button');
+    this.addAction(skipVersionButton, this.skipVersion.bind(this), 'click');
+    this.controlPanel.appendChild(skipVersionButton);
+
+    this.appendNotNowButton();
+
+    const downloadButton = this.createButton(
+      nls.localize('arduino/ide-updater/downloadButton', 'Download')
+    );
+    this.addAction(downloadButton, this.startDownload.bind(this), 'click');
+    this.controlPanel.appendChild(downloadButton);
+    downloadButton.focus();
+  }
+
+  private appendInstallButtons(): void {
+    this.clearButtons();
+    this.appendNotNowButton();
+
+    const closeAndInstallButton = this.createButton(
+      nls.localize(
+        'arduino/ide-updater/closeAndInstallButton',
+        'Close and Install'
+      )
+    );
+    this.addAction(
+      closeAndInstallButton,
+      this.updater.quitAndInstall.bind(this),
+      'click'
+    );
+    this.controlPanel.appendChild(closeAndInstallButton);
+    closeAndInstallButton.focus();
+  }
+
+  private appendErrorButtons(): void {
+    this.clearButtons();
+    this.appendNotNowButton();
+
+    const goToDownloadPageButton = this.createButton(
+      nls.localize('arduino/ide-updater/goToDownloadButton', 'Go To Download')
+    );
+    this.addAction(
+      goToDownloadPageButton,
+      this.openDownloadPage.bind(this),
+      'click'
+    );
+    this.controlPanel.appendChild(goToDownloadPageButton);
+    goToDownloadPageButton.focus();
+  }
+
+  private openDownloadPage() {
+    this.windowService.openNewWindow(DOWNLOAD_PAGE_URL, { external: true });
```

Comment:
If the `DOWNLOAD_PAGE_URL` is not used elsewhere, it's safer to have it as a string URL so that no magic can happen at runtime. This is a security consideration.

Reply:
Here is the electron recommendation: https://www.electronjs.org/docs/latest/tutorial/security#15-do-not-use-shellopenexternal-with-untrusted-content.

I know that the code you modified is not calling electron APIs, but I'd propose following their guidelines. What do you think?

Reply:
I never thought about it, thanks for this remark.

What I read on Electron documentation is:

>  When openExternal is used with untrusted content, it can be leveraged to execute arbitrary commands.

It mentions **untrusted content** to execute **arbitrary commands**, and the example refers to **user controlled data**. I'm not a security expert, but what I understand from the documentation is that using `openExternal` with a parameter that might change after some user action is risky. In this case, `DOWNLOAD_PAGE_URL` is a constant and it's not possible to change it at runtime, so there shouldn't be this risk, right?

Reply:
> I never thought about it, thanks for this remark.

Me neither, until I read about this a few weeks ago.

> so there shouldn't be this risk, right?

It's up to you. The snippets with the `bad` and `good` examples from the electron docs are straightforward. If you have concerns, ask somebody from the security team for the final decision.

Reply:
Since the value is hard-coded, I believe it's pretty safe, so I wouldn't change it in this PR.

Also, we're doing something similar in [other](https://github.com/arduino/arduino-ide/blob/main/arduino-ide-extension/src/browser/widgets/cloud-sketchbook/cloud-sketchbook-tree-widget.tsx#L142-L144) [places](https://github.com/arduino/arduino-ide/blob/main/arduino-ide-extension/src/browser/contributions/survey-notification.ts#L61-L64), so if we decide to apply your suggestion as a security practice, we should apply it across all the project.

Reply:
As you wish. 👍